fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexingHeartbeatTest.java

```diff
@@ -0,0 +1,182 @@
+import com.apple.foundationdb.record.metadata.IndexTypes;
+import com.apple.foundationdb.record.metadata.expressions.EmptyKeyExpression;
+import com.apple.foundationdb.tuple.Tuple;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Verify indexing heartbeat activity (query & clear).
+ */
+class OnlineIndexingHeartbeatTest extends OnlineIndexerTest {
+
+    @Test
+    void testHeartbeatLowLevel() {
+        List<Index> indexes = new ArrayList<>();
+        indexes.add(new Index("indexA", field("num_value_2"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS));
+        indexes.add(new Index("indexB", field("num_value_3_indexed"), IndexTypes.VALUE));
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+
+        final int count = 10;
+        IndexingHeartbeat[] heartbeats = new IndexingHeartbeat[count];
+        for (int i = 0; i < count; i++) {
+            heartbeats[i] = new IndexingHeartbeat(UUID.randomUUID(), IndexBuildProto.IndexBuildIndexingStamp.Method.BY_INDEX, 100 + i);
+        }
+
+        openSimpleMetaData(hook);
+        try (FDBRecordContext context = openContext()) {
+            for (var heartbeat : heartbeats) {
+                heartbeat.updateHeartbeat(recordStore, indexes.get(0));
+                heartbeat.updateHeartbeat(recordStore, indexes.get(1));
+            }
+            context.commit();
+        }
+
+        // Verify query/clear operation
+        try (OnlineIndexer indexer = newIndexerBuilder(indexes.get(0)).build()) {
+            // Query, unlimited
+            Map<UUID, IndexBuildProto.IndexBuildHeartbeat> queried = indexer.getIndexingHeartbeats(0);
+            Assertions.assertThat(queried).hasSize(count);
+            Assertions.assertThat(queried.keySet())
+                    .containsExactlyInAnyOrderElementsOf(Arrays.stream(heartbeats).map(ht -> ht.indexerId).collect(Collectors.toList()));
+
+            // Query, partial
+            queried = indexer.getIndexingHeartbeats(5);
+            Assertions.assertThat(queried).hasSize(5);
+
+            // clear, partial
+            int countDeleted = indexer.clearIndexingHeartbeats(0, 7);
+            Assertions.assertThat(countDeleted).isEqualTo(7);
+            queried = indexer.getIndexingHeartbeats(5);
+            Assertions.assertThat(queried).hasSize(3);
+        }
+
+        // Verify that the previous clear does not affect other index
+        try (OnlineIndexer indexer = newIndexerBuilder(indexes.get(1)).build()) {
```

Comment:
There should be low-level tests of updating leases, and ensuring that the heartbeat stays non-expired.
Also, that it can steal an expired lease. And that the original heartbeat will fail it lost its heartbeat. I think it would only fail if the stealing process is still running so perhaps covering it trying to continue while the stealer is still active, and if that has completed.